yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -825,6 +813,9 @@ private[spark] class Client(
    * throw an appropriate SparkException.
    */
   def run(): Unit = {
+    // Cleanup staging director as some appStagingDir can not be deleted when job is failed or
+    // killed, please see SPARK-7705 for details.
+    cleanupStagingDir()
```

Comment:
1. Cleanup old application staging directory before submit current application.
2. Yes, if called on run on yarn-client will not work, yarn-cluster is ok, so called on submitApplication().